server/BudgetTracker.Business/Api/BudgetApi.cs

```diff
@@ -153,5 +149,38 @@ public async Task<ApiResponse> GetRootBudgets(ApiRequest request)
             response = new ApiResponse(responseData);
             return response;
         }
-    }    
+
+        public Task<ApiResponse> FetchBudgetTree(ApiRequest request)
+        {
+            User user = await Authenticate(request);
+            ApiResponse response;
+
+            Guid rootBudgetId = request.Arguments<FetchBudgetTreeArgumentsApiContract>();
+
+            try
+            {
+                Budget rootBudget = GetBudgetOrErrorIfUnauthorized(rootBudgetId, user.Id);
+
+                await _budgetRepository.LoadBudgetTree(rootBudget);
+            }
+            catch (RepositoryException ex)
+            {
+                response.Error = ex.Message;
+            }
+
+            return response;
+        }
+
+        private async Task<Budget> GetBudgetOrErrorIfUnauthorized(Guid budgetId, Guid userId)
+        {
+            Budget retrievedBudget = await _budgetRepository.GetBudget(budgetId);
+
+            if (retrievedBudget.Owner.Id != userId)
+            {
+                throw new RepositoryException("The authenticated user does not have permission to view this budget.");
```

Comment:
Is this really a Repository Exception? I think I might have had it like this in my implementation however, when I see it here it makes me think it should be something different.

Reply:
Yeah I just copied and pasted from yours but I can change it.